### PR TITLE
Feature/expires at subscription service

### PR DESCRIPTION
**Pull request by PoulomiNandy**
## Proposed changes

Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
If it fixes a bug or resolves a feature request, be sure to link to that issue.

## Types of changes

What types of changes does your code introduce to the project: _Put an `x` in the boxes that apply_

-   [ ] Bugfix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of
them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before
merging your code._

-   [ ] I have read the [CONTRIBUTING](https://github.com/stellio-hub/stellio-context-broker/blob/develop/CONTRIBUTING.md) doc
-   [x] I have signed the [CLA](https://fiware.github.io/contribution-requirements/individual-cla.pdf)
-   [ ] I have added tests that prove my fix is effective or that my feature works
-   [ ] I have added necessary documentation (if appropriate)
-   [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you
did and what alternatives you considered, etc...

### CODE REVIEW

File: search-service/src/main/kotlin/com/egm/stellio/search/config/ApplicationProperties.kt [hunk 1]
```diff
@@ -6,7 +6,7 @@ import java.net.URI
 
 @ConstructorBinding
 @ConfigurationProperties("application")
-data class ApplicationProperties(
+data class dockerApplicationProperties(
```
Comment (bobeal) [inline search-service/src/main/kotlin/com/egm/stellio/search/config/ApplicationProperties.kt:9]: I don't think this was expected?

File: search-service/src/main/resources/application.properties [hunk 2]
```diff
@@ -29,6 +29,10 @@ management.endpoint.health.enabled = true
 
 spring.mvc.log-request-details = true
 
+# Pagination config for query resources endpoints
+application.pagination.limit-default = 30
+application.pagination.limit-max = 100
+
```
Comment (bobeal) [inline search-service/src/main/resources/application.properties:35]: You should not have these changes in this PR

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/model/Subscription.kt [hunk 3]
```diff
@@ -83,7 +84,7 @@ private fun serializeWithoutSysAttrs(input: Any) =
         Subscription::class,
         SysAttrsMixinFilter::class,
         "sysAttrs",
-        setOf("createdAt", "modifiedAt")
+        setOf("createdAt", "modifiedAt", "expiresAt")
```
Comment (bobeal) [inline subscription-service/src/main/kotlin/com/egm/stellio/subscription/model/Subscription.kt:87]: `expiresAt` is not one of the sysAttrs parameter (see 6.3.11)

File: search-service/src/main/resources/application.properties (hunk 2 above)
Comment (PoulomiNandy) [reply search-service/src/main/resources/application.properties:35]: Strange I really do not know why this is showing in this PR. I didn't even touch search service.

File: subscription-service/src/main/resources/db/migration/V0_12__add_expires_at_to_subscription_table.sql [hunk 4]
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE subscription
+ADD expires_at date;
```
Comment (bobeal) [inline subscription-service/src/main/resources/db/migration/V0_12__add_expires_at_to_subscription_table.sql:2]: for dates columns, we use TIMESTAMPTZ type (in NGSI-LD, dates are always represented at UTC time)

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/service/SubscriptionService.kt [hunk 5]
```diff
@@ -404,7 +408,8 @@ class SubscriptionService(
             SELECT subscription.id as sub_id, subscription.type as sub_type, name, description, q,
                    notif_attributes, notif_format, endpoint_uri, endpoint_accept, times_sent, endpoint_info
             FROM subscription 
-            WHERE is_active 
+            WHERE is_active
+            AND (expires_at is null OR expires_at >= now())
```
Comment (bobeal) [inline subscription-service/src/main/kotlin/com/egm/stellio/subscription/service/SubscriptionService.kt:412]: be careful with timezones when comparing the two dates. `expires_at` is expected to be at UTC, `now()` is not always (you can use `Instant.now().atZone(ZoneOffset.UTC)` instead)

File: subscription-service/src/test/kotlin/com/egm/stellio/subscription/service/SubscriptionServiceTests.kt [hunk 6]
```diff
@@ -142,11 +145,52 @@ class SubscriptionServiceTests : WithTimescaleContainer {
...
             ),
-            isActive = false
+            isActive = false,
+            expiresAt = ZonedDateTime.parse("2012-08-12T08:33:38Z")
         )
         subscription6Id = createSubscription(subscription)
     }
 
+    @Test
+    fun `it should not retrive an expired subscription matching an id`() {
```
Comment (bobeal) [inline subscription-service/src/test/kotlin/com/egm/stellio/subscription/service/SubscriptionServiceTests.kt:155]:
```suggestion
    fun `it should not retrieve an expired subscription matching an id`() {
```

File: subscription-service/src/test/kotlin/com/egm/stellio/subscription/service/SubscriptionServiceTests.kt [hunk 7]
```diff
@@ -142,11 +145,52 @@ class SubscriptionServiceTests : WithTimescaleContainer {
...
+                "incoming"
+            )
+
+        StepVerifier.create(persistedSubscription)
+            .expectNextCount(0)
+            .verifyComplete()
+    }
+
+    @Test
+    fun `it should retrive an subscription matching an id when expired date is not given`() {
```
Comment (bobeal) [inline subscription-service/src/test/kotlin/com/egm/stellio/subscription/service/SubscriptionServiceTests.kt:169]:
```suggestion
    fun `it should retrieve a subscription matching an id when expired date is not given`() {
```

File: subscription-service/src/test/kotlin/com/egm/stellio/subscription/service/SubscriptionServiceTests.kt [hunk 8]
```diff
@@ -142,11 +145,52 @@ class SubscriptionServiceTests : WithTimescaleContainer {
...
+                "smartDoor",
+                "incoming"
+            )
+        StepVerifier.create(persistedSubscription)
+            .expectNextCount(1L)
+            .verifyComplete()
+    }
+
+    @Test
+    fun `it should retrive an subscription matching an id when expired date is in the future`() {
```
Comment (bobeal) [inline subscription-service/src/test/kotlin/com/egm/stellio/subscription/service/SubscriptionServiceTests.kt:182]:
```suggestion
    fun `it should retrieve a subscription matching an id when expired date is in the future`() {
```

File: subscription-service/src/test/kotlin/com/egm/stellio/subscription/service/SubscriptionServiceTests.kt [hunk 9]
```diff
@@ -87,7 +88,8 @@ class SubscriptionServiceTests : WithTimescaleContainer {
             entities = setOf(
                 EntityInfo(id = null, idPattern = null, type = "Beekeeper"),
                 EntityInfo(id = "urn:ngsi-ld:Beehive:1234567890".toUri(), idPattern = null, type = "Beehive")
-            )
+            ),
+            expiresAt = ZonedDateTime.parse("2050-12-12T08:33:38Z")
```
Comment (bobeal) [inline subscription-service/src/test/kotlin/com/egm/stellio/subscription/service/SubscriptionServiceTests.kt:92]: I know that 2050 is far away in the future but you can do something more robust with, for instance: `Instant.now().atZone(ZoneOffset.UTC).plusDays(1)`

**Review by bobeal — APPROVED** [review]

**Review by bobeal — APPROVED** [review]

**Pull request merged**